### ISSUE TITLE
Crawler hangs on non-html URL requests

### ISSUE THREAD

**Issue by lausycampari** [yujiosaka/headless-chrome-crawler#136]
Currently, when attempting to crawl a non-html page such as a PDF, the crawler seems to hang for a while, and eventually fails silently without giving any error. How should we go about handling non-html files? My thought was to use `preRequest` function to modify some options after checking the mime-type, like setting jQuery to false, etc.; but I've not been able to get anything to work for non-html content types. How would you recommend accommodating this functionality? 

Here is a sample to reproduce this, which shows that the URL is queued, but eventually the crawler exits with neither onSuccess nor onError being called. Ideally, it would be nice to have some logic to intercept non-html requests and route them to use Apache tika or something similar.

```
const HCCrawler = require('headless-chrome-crawler');

HCCrawler.launch({
  args: ['--disable-web-security'],
  jQuery: false,
  onSuccess: (result => {
    console.log(`crawled: ${result.response.url}`);
  }),
  onError: (error => {
    console.error(error);
  })
})
.then(crawler => {
  crawler.queue(['https://pdfs.semanticscholar.org/8f86/834ae39f46447fd588b5817d6d9171f518e6.pdf']);
  crawler.queueSize()
    .then(size => {
      console.log('%s items in cache', size);
    });
  crawler.onIdle()
    .then(() => crawler.close());
});
```

**Comment by lausycampari** [comment]
I managed to track this down to the chromium project [here](https://bugs.chromium.org/p/chromium/issues/detail?id=761295) - Apparently you must have `headless` set to `false`, and then the PDF will be fetched successfully. But that still leaves the problem of the text extraction, and link extraction. So it would just be a matter of detecting when a resource is non-html so that `headless` can be set to `false` before the request is made, and `jQuery` forcefully overriden, since that wouldn't be of any use on a pdf. Maybe something like [this](https://github.com/GoogleChrome/puppeteer/issues/610#issuecomment-340160025) could be integrated?

**Comment by yujiosaka** [comment]
@lausycampari 
It seems that the bug is fixed after updating puppeteer v1.5.0.
None of the release note seems related to this issue, so rolling Chromium to `69.0.3452.0 (r564778)` happened to solve the issue.

This is what I get when I output `result` in `onSuccess`:
```json
{ response:
   { ok: true,
     url: 'https://www.semanticscholar.org/paper/Actin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer/8f86834ae39f46447fd588b5817d6d9171f518e6',
     status: 200,
     headers:
      { status: '200',
        date: 'Sat, 09 Jun 2018 07:32:00 GMT',
        'content-type': 'text/html; charset=utf-8',
        server: 'nginx',
        vary: 'Accept-Encoding',
        'strict-transport-security': 'max-age=31536000; includeSubdomains',
        'cache-control': 'max-age=0, no-cache, no-store',
        expires: '-1',
        'x-frame-options': 'SAMEORIGIN',
        'content-encoding': 'gzip' } },
  redirectChain:
   [ { url: 'https://pdfs.semanticscholar.org/8f86/834ae39f46447fd588b5817d6d9171f518e6.pdf',
       headers: [Object] },
     { url: 'https://www.semanticscholar.org/paper/8f86834ae39f46447fd588b5817d6d9171f518e6?p2df',
       headers: [Object] } ],
  result: null,
  screenshot: null,
  cookies:
   [ { name: '_hp2_props.2424575119',
       value: '%7B%22experiment%3Aalert%22%3A%22visible%22%2C%22experiment%3Aautocomplete%22%3A%22fresh%22%2C%22experiment%3Apdp_search_callout%22%3Anull%2C%22experiment%3Asimilar_papers_meta%22%3Anull%2C%22experiment%3Areader_launch%22%3A%22control%22%7D',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1591601522,
       size: 261,
       httpOnly: false,
       secure: false,
       session: false },
     { name: '_hp2_ses_props.2424575119',
       value: '%7B%22ts%22%3A1528529521571%2C%22d%22%3A%22www.semanticscholar.org%22%2C%22h%22%3A%22%2Fpaper%2FActin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer%2F8f86834ae39f46447fd588b5817d6d9171f518e6%22%7D',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1528531321,
       size: 232,
       httpOnly: false,
       secure: false,
       session: false },
     { name: '_gat',
       value: '1',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1528529582,
       size: 5,
       httpOnly: false,
       secure: false,
       session: false },
     { name: '_gid',
       value: 'GA1.2.1179173506.1528529521',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1528615921,
       size: 31,
       httpOnly: false,
       secure: false,
       session: false },
     { name: '_hp2_id.2424575119',
       value: '%7B%22userId%22%3A%225969496914199408%22%2C%22pageviewId%22%3A%228737469070792071%22%2C%22sessionId%22%3A%223519485054172357%22%2C%22identity%22%3Anull%2C%22trackerVersion%22%3A%224.0%22%7D',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1591601521,
       size: 207,
       httpOnly: false,
       secure: false,
       session: false },
     { name: '_ga',
       value: 'GA1.2.653010433.1528529521',
       domain: '.semanticscholar.org',
       path: '/',
       expires: 1591601521,
       size: 29,
       httpOnly: false,
       secure: false,
       session: false },
     { name: 'tid',
       value: 'rBIABlsbgnA+ZwAIGUOcAg==',
       domain: 'www.semanticscholar.org',
       path: '/',
       expires: 2145916555.175644,
       size: 27,
       httpOnly: false,
       secure: false,
       session: false } ],
  links:
   [ 'https://www.semanticscholar.org/',
     'https://www.semanticscholar.org/faq',
     'https://www.semanticscholar.org/author/Timothy-J-Mitchison/4294579',
     'https://www.semanticscholar.org/author/L.-P.-Cramer/34980782',
     'https://www.ncbi.nlm.nih.gov/pubmed/8608590',
     'http://doi.org/10.1016/S0092-8674%2800%2981281-7',
     'https://www.semanticscholar.org/topic/cell-motility/22083',
     'https://www.semanticscholar.org/topic/Blood-Cells/10007',
     'https://www.semanticscholar.org/topic/Traction/15827',
     'https://www.semanticscholar.org/topic/Dental-Retraction-(procedure)/1679',
     'https://www.semanticscholar.org/paper/Actin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer/8f86834ae39f46447fd588b5817d6d9171f518e6/figure/0',
     'https://www.semanticscholar.org/paper/Actin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer/8f86834ae39f46447fd588b5817d6d9171f518e6/figure/1',
     'https://www.semanticscholar.org/paper/Actin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer/8f86834ae39f46447fd588b5817d6d9171f518e6/figure/2',
     'https://www.semanticscholar.org/paper/Actin-Based-Cell-Motility-and-Cell-Locomotion-Mitchison-Cramer/8f86834ae39f46447fd588b5817d6d9171f518e6/figure/3',
     'https://www.semanticscholar.org/paper/Stability%2C-convergence%2C-and-sensitivity-analysis-of-Sfakianakis-Brunk/ac2349376ec47ab906cf614a54c6ead462e0da9a',
     'https://www.semanticscholar.org/author/Nikolaos-Sfakianakis/1691407',
     'https://www.semanticscholar.org/author/Aaron-Brunk/8235887',
     'https://www.semanticscholar.org/paper/Polymeric-Biomaterials-for-Tissue-Regeneration-Gao-Chong/a359e2e1fefc3bcb653e684dd8b10e745a3dcb51',
     'https://www.semanticscholar.org/author/Changyou-Gao/2220650',
     'https://www.semanticscholar.org/author/Mark-Seow-Khoon-Chong/35156243',
     'https://www.semanticscholar.org/author/Changsheng-Liu/5520017',
     'https://www.semanticscholar.org/author/Xiaowen-Zheng/35201532',
     'https://www.semanticscholar.org/paper/Collective-epithelial-migration-and-cell-drive-Ewald-Brenot/6e3ad0213d2667f714b761d9b153129bc276c45e',
     'https://www.semanticscholar.org/author/Andrew-J-Ewald/1851633',
     'https://www.semanticscholar.org/author/Audrey-Brenot/5184123',
     'https://www.semanticscholar.org/author/Myhanh-Duong/11774723',
     'https://www.semanticscholar.org/author/Bianca-S-Chan/39869914',
     'https://www.semanticscholar.org/author/Zena-Werb/4232463',
     'https://www.semanticscholar.org/paper/A-Minimal-Model-of-Locomotion-Applied-to-the-Steady-MOGILNER/23d6aa6a3947933d421d1499921e21fb22f6cc24',
     'https://www.semanticscholar.org/author/A.-MOGILNER/36739955',
     'https://www.semanticscholar.org/paper/A-Model-for-Myosin-Powered-Dynamic-Contraction-of-Mogilner-Svitkina/e6386730a080d41033f49074ec13ed5bbbabdaaa',
     'https://www.semanticscholar.org/author/A.-Mogilner/1759829',
     'https://www.semanticscholar.org/author/T.-Svitkina/6081874',
     'https://www.semanticscholar.org/author/A.-Verkhovsky/2260520',
     'https://www.semanticscholar.org/paper/Localized-RhoA-activation-as-a-requirement-for-the-Kurokawa-Matsuda/424eca0ed12fbab16513aa2992a1b5e73a12d8ed',
     'https://www.semanticscholar.org/author/Kazuo-Kurokawa/5537645',
     'https://www.semanticscholar.org/author/Michiyuki-Matsuda/3114551',
     'https://www.semanticscholar.org/paper/Dopamine-Increases-CD14%2BCD16%2B-Monocyte-across-the-Calderon-Williams/7a76a3472ee0d933b926520c30b8776506a608b5',
     'https://www.semanticscholar.org/author/Tina-M.-Calderon/6214211',
     'https://www.semanticscholar.org/author/Dionna-W.-Williams/35087809',
     'https://www.semanticscholar.org/author/Joan-W.-Berman/2394279',
     'https://www.semanticscholar.org/paper/Redox-regulation-of-the-actin-cytoskeleton-and-its-Xu-Huff/c179b7dc1d4d095235a305adaafd298ff78f539a',
     'https://www.semanticscholar.org/author/Qian-Xu/1784783',
     'https://www.semanticscholar.org/author/Lauren-P-Huff/40659446',
     'https://www.semanticscholar.org/author/Masakazu-Fujii/2458777',
     'https://www.semanticscholar.org/author/Kathy-K-Griendling/2593114',
     'https://www.semanticscholar.org/paper/Amebiasis-Nozaki-Bhattacharya/1724646e42a1236640e298d5725fcb5b871cc571',
     'https://www.semanticscholar.org/author/Alok-Bhattacharya/1714693',
     'https://www.semanticscholar.org/author/Tomoyoshi-Nozaki/38504813',
     'https://www.semanticscholar.org/paper/Response-of-an-actin-filament-network-model-under-a-Kang-Steward/73f4c7c4e9dd710dd7dd47b17cf347fd0d3bdfdd',
     'https://www.semanticscholar.org/author/John-Kang/4547002',
     'https://www.semanticscholar.org/author/Robert-L-Steward/38759479',
     'https://www.semanticscholar.org/author/YongTae-Kim/40219382',
     'https://www.semanticscholar.org/author/Russell-S-Schwartz/40461815',
     'https://www.semanticscholar.org/author/Philip-R-LeDuc/2094854',
     'https://www.semanticscholar.org/author/Kathleen-M-Puskar/35685268',
     'https://www.semanticscholar.org/paper/Actin-microfilament-dynamics-in-locomoting-cells-Theriot-Mitchison/ce77d1f9886952cd9d83761659887da9744706ee',
     'https://www.semanticscholar.org/author/Julie-A.-Theriot/2116997',
     'https://www.semanticscholar.org/author/Timothy-J.-Mitchison/4294579',
     'https://www.semanticscholar.org/paper/The-organization-of-myosin-and-actin-in-rapid-nerve-Bridgman-Dailey/d73f99d65cb887baca00c7aa124eda1202149284',
     'https://www.semanticscholar.org/author/P-C-Bridgman/33607536',
     'https://www.semanticscholar.org/author/M-E-Dailey/4995836',
     'https://www.semanticscholar.org/paper/The-actin-cytoskeleton.-Small/2aae0a699ae5ee44f7895fadfea5517cf7b3eb08',
     'https://www.semanticscholar.org/author/J-V-Small/1815831',
     'https://www.semanticscholar.org/paper/Actin-based-bacterial-motility%3A-towards-a-of-the-Lasa-Cossart/b5189b792409e0b5392da9b38bc7b6e1473a81d0',
     'https://www.semanticscholar.org/author/I-Lasa/32542011',
     'https://www.semanticscholar.org/author/P-Cossart/2347622',
     'https://www.semanticscholar.org/paper/Actin-filament-organization-in-the-fish-keratocyte-Small-Herzog/37de8bfa9edfc2c20bffa53ff680510fe2c7c6e6',
     'https://www.semanticscholar.org/author/M-Herzog/2664799',
     'https://www.semanticscholar.org/author/K-Anderson/1714430',
     'https://www.semanticscholar.org/paper/Identification-of-Novel-Graded-Polarity-Actin-in-of-Cramer-Siebert/44cbe6616d54c10d047c60cbce164834b250e6a1',
     'https://www.semanticscholar.org/paper/Analysis-of-the-Actin%E2%80%93Myosin-II-System-in-Fish-of-Svitkina-Verkhovsky/0b4deafb6f1a7deb58ebee63ab00910d4ddf466c',
     'https://www.semanticscholar.org/paper/Localized-Depolymerization-of-the-Major-Sperm-with-Italiano-Stewart/4d08e4fe59b690e26dc92aae1bd821a3c5775d20',
     'https://www.semanticscholar.org/paper/Myosin-is-involved-in-postmitotic-cell-spreading-Cramer-Mitchison/23977421ab98a2db839344ec930cc7b22d1002da',
     'https://www.semanticscholar.org/paper/Cell-Migration%3A-A-Physically-Integrated-Molecular-Lauffenburger-Horwitz/c9e574b26bee6a4a3489c88316fe89115d8bb69b',
     'https://www.semanticscholar.org/search?q=cell%20motility&sort=relevance',
     'http://allenai.org/',
     'https://www.semanticscholar.org/terms-of-service',
     'https://www.semanticscholar.org/privacy-policy' ],
  options:
   { maxDepth: 1,
     priority: 0,
     delay: 0,
     retryCount: 3,
     retryDelay: 10000,
     timeout: 30000,
     jQuery: false,
     browserCache: true,
     skipDuplicates: true,
     depthPriority: true,
     obeyRobotsTxt: true,
     followSitemapXml: false,
     skipRequestedRedirect: false,
     cookies: null,
     screenshot: null,
     viewport: null,
     url: 'https://pdfs.semanticscholar.org/8f86/834ae39f46447fd588b5817d6d9171f518e6.pdf' },
  depth: 1 }
```

**Comment by yujiosaka** [comment]
Nope, it wasn't fixed.
The shared url started working suddenly, but it was just the behavior of the url has been changed: redirect to a standard HTML page.

I tried [another URL](http://unec.edu.az/application/uploads/2014/12/pdf-sample.pdf) and it fails.
But it's not bad because accessing to the URL causes `net::ERR_ABORTED` and you can catch it.

I should keep this issue closed now because there is little we can do for it.
